platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfiguratorImpl.java

```diff
@@ -0,0 +1,465 @@
+ * When the {@link #commit()} operation completes - either successfully, with a successful rollback,
+ * or with a failure to rollback - it returns a {@link OperationReport} of the outcome. In the case of
+ * rollback failures, callers of this class should inform users of those failures so they may manually
+ * intercede.
+ * <p>
+ * This class does not guarantee that it can reliably rollback changes in the case of failure. It
+ * makes a best-effort to revert changes and reports the outcome.
+ * <p>
+ * To use this class, first instantiate then invoke the various methods for feature, bundle, config,
+ * etc. updates in the order they should be applied. When all have been completed, call the
+ * {@link #commit()} method to write the changes to the system. The resulting {@link OperationReport}
+ * will have the outcome.
+ */
+public class ConfiguratorImpl implements Configurator, ConfigReader {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfiguratorImpl.class);
+
+    private final Map<UUID, Operation> configHandlers = new LinkedHashMap<>();
+
+    /**
+     * Sequentially invokes all the {@link Operation}s, committing their changes. If a failure
+     * occurs during the processing, a rollback is attempted of those handlers that had already been
+     * committed.
+     * <p>
+     * In the case of a successful commit, the {@link SecurityLogger} will be invoked to log changes.
+     *
+     * @param auditMessage In the case of a successful commit, the message to pass to the
+     *                     {@link SecurityLogger}
+     * @param auditParams  In the case of a successful commit, the optional parameters to pass to the
+     *                     {@link SecurityLogger} to be interpolated into the message
+     * @return report of the commit status, whether successful, successfully rolled back, or partially
+     * rolled back with errors
+     */
+    public OperationReport commit(String auditMessage, String... auditParams) {
+        OperationReport report = commit();
+        if (report.hasTransactionSucceeded()) {
+            SecurityLogger.audit(auditMessage, (Object[]) auditParams);
+        }
+
+        return report;
+    }
+
+    /**
+     * Sequentially invokes all the {@link Operation}s, committing their changes. If a failure
+     * occurs during the processing, a rollback is attempted of those handlers that had already been
+     * committed.
+     *
+     * @return report of the commit status, whether successful, successfully rolled back, or partially
+     * rolled back with errors
+     */
+    public OperationReport commit() {
+        OperationReport configReport = new OperationReportImpl();
+        for (Map.Entry<UUID, Operation> row : configHandlers.entrySet()) {
+            try {
+                Object commitResult = row.getValue()
+                        .commit();
+                if (commitResult instanceof String) {
```

Comment:
This smells of tight coupling and leaking implementation details that a String response is a passing managed service operation.